[Security] Update custom authenticator docs to include identifier normalization

Spomky opened the pull request:
Fixes #20632

Improve the documentation for custom authenticators by detailing the concept of user identifier normalization. Introduced the `NormalizedUserBadge` example to demonstrate how to simplify and standardize identifiers.

xabbuh commented on security/custom_authenticator.rst:
```diff
@@ -209,13 +209,20 @@ requires a user and some sort of "credentials" (e.g. a password).
+~~~~~~~~~~~~~~~
+
+The user identifier is a unique string that identifies the user. It is used
+ to load the user using :ref:`the user provider <security-user-providers>`.
```
```suggestion
to load the user using :ref:`the user provider <security-user-providers>`.
```

alamirault commented on security/custom_authenticator.rst:
```diff
@@ -255,6 +262,89 @@ using :ref:`the user provider <security-user-providers>`::
+or "JOHN.DOE" refer to the same user.
+Normalization can include converting the identifier to lowercase
+and trimming unnecessary spaces.
+You can optionally pass a user identifier normalizer as third argument to the
+``UserBadge``. This callable receives the ``$userIdentifier``
+and must return a normalized user identifier as a string.
+
+.. versionadded:: 7.3
```
New section is not correctly rendered:

![image](https://github.com/user-attachments/assets/d5560800-f910-441b-a142-c98e4f30e1bb)

I don't know how to fix it

Spomky replied:
Many thanks.
I will take time to update this section.

Spomky replied:
It should be fixed with the last commit

alamirault commented on security/custom_authenticator.rst:
```diff
@@ -255,6 +262,89 @@ using :ref:`the user provider <security-user-providers>`::
+        }
+    }
+
+.. note::
+
+    Similarly, Google normalizes email addresses so that "john.doe", "j.hon.d.oe",
+    and "johndoe" all correspond to the same account.
+    This involves removing dots and converting the email address to lowercase
+    (though normalization specifics depend on your use case).
```
With the other note below, I'm not sure this one is needed

Spomky replied:
I believe both notes provide valuable context and would prefer to keep tthem:
* The first shows a well-known real-world example (Google).
* The second one highlights enterprise-specific formats where normalization is important.

Spomky committed:
Introduce user identifier normalization for enhanced consistency

Added support for normalizing user identifiers in Symfony 7.3 to ensure consistent comparison and storage. Updated documentation with implementation examples, highlighting best practices like converting identifiers to lowercase and handling various formats. This improvement reduces the risk of duplicates and improves reliability in user authentication.

alamirault commented on security/custom_authenticator.rst:
```diff
@@ -255,6 +262,91 @@ using :ref:`the user provider <security-user-providers>`::
+
+    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+    use Symfony\Component\String\UnicodeString;
+    use function Symfony\Component\String\u;
+
+    final class NormalizedUserBadge extends UserBadge
+    {
+        public function __construct(string $identifier)
+        {
+            $callback = static fn (string $identifier) => u($identifier)->normalize(UnicodeString::NFKC)->ascii()->lower()->toString();
+
+            parent::__construct($identifier, null, $callback);
+        }
+    }
+
+    // src/Security/PasswordAuthenticator.php
+    namespace App\Security;
+
+    final class PasswordAuthenticator extends AbstractLoginFormAuthenticator
+    {
+        // Simplified for brievety
+        public function authenticate(Request $request): Passport
+        {
+            $username = (string) $request->request->get('username', '');
+            $password = (string) $request->request->get('password', '');
+
+            $request->getSession()
+                ->set(SecurityRequestAttributes::LAST_USERNAME, $username);
+
+            return new Passport(
+                new NormalizedUserBadge($username),
+                new PasswordCredentials($password),
+                [
+                    //All other useful badges
+                ]
+            );
+        }
+    }
+
+.. note::
```
- Php code was not rendered. I added `::` and  `.. code-block:: php` (Not 100% sure it's the best way)
- I replace `(string) $request->request->get()` by `$request->request->getString`
```suggestion
For instance, the example below uses a normalizer that converts usernames to a normalized, ASCII-only, lowercase format,
suitable for consistent comparison and storage::

    // src/Security/NormalizedUserBadge.php
    namespace App\Security;

    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
    use Symfony\Component\String\UnicodeString;
    use function Symfony\Component\String\u;

    final class NormalizedUserBadge extends UserBadge
    {
        public function __construct(string $identifier)
        {
            $callback = static fn (string $identifier) => u($identifier)->normalize(UnicodeString::NFKC)->ascii()->lower()->toString();

            parent::__construct($identifier, null, $callback);
        }
    }

.. code-block:: php

    // src/Security/PasswordAuthenticator.php
    namespace App\Security;

    final class PasswordAuthenticator extends AbstractLoginFormAuthenticator
    {
        // Simplified for brievety
        public function authenticate(Request $request): Passport
        {
            $username = $request->request->getString('username');
            $password = $request->request->getString('password');

            $request->getSession()
                ->set(SecurityRequestAttributes::LAST_USERNAME, $username);

            return new Passport(
                new NormalizedUserBadge($username),
                new PasswordCredentials($password),
                [
                    //All other useful badges
                ]
            );
        }
    }

.. note::
```

javiereguiluz commented:
Thanks Florent for this contribution ... and thanks reviewers for your nice review!
